### DISCUSSION TITLE
Built-in LSP support?

### DISCUSSION THREAD

**Discussion by bbatsov** [rubocop/rubocop#11686] [Ideas]
I've noticed that StandardRB has added an LSP server recently (see https://blog.testdouble.com/posts/2023-02-16-its-official-the-standard-ruby-vscode-extension/) and I've been wondering if that'd be a reasonable direction for us to take as well. Obviously RuboCop can't provide support for everything in the LSP API but formatting and linting is probably plenty for some people. 

On the other other hand I think some fully-fledged Ruby LSP servers are using RuboCop internally anyways. Admittedly I don't follow the trends in this department that much.

**Comment by cvx** [comment]
(A note for people coming from e.g. Dependabot PRs: [1.48.1 release notes](https://github.com/rubocop/rubocop/releases/tag/v1.48.1) incorrectly linked "Fix `Lint/Syntax` behavior when `Enabled: false` of `Lint` department." to this discussion. The correct PR number is #11689)

**Reply by koic** [reply]
I've fixed the number. Thanks!

**Comment by mike-burns** [comment]
This was done for performance reasons.

Solargraph formats code by running Rubocop from scratch every time. This is typically done on file write, and it has to be synchronous: the editor can't allow input until the formatting has finished (otherwise it'll then have to reconcile the edits against a file that has changed underneath it).

So a synchronous format on file write is great unless the formatter is slow. With Solargraph using Rubocop, I'm seeing writes that take 3 seconds or more.

Standardrb's `--lsp` flag makes it run as a daemon, and this improves the performance to the point where it's not noticeable.

Note that you can run multiple LSP servers concurrently. I can imagine a world where Solargraph provides tab completion, refactoring, etc., and Rubocop provides formatting and linting.

**Comment by koic** [comment]
FYI, I opened PR https://github.com/rubocop/rubocop/pull/11926 to support built-in LSP.

**Comment by koic** [comment]
Built-in RuboCop LSP server is now supported on RuboCop 1.53+.
https://docs.rubocop.org/rubocop/1.53/usage/lsp.html

If you are a VS Code user, you can use vscode-rubocop which is a VS Code extension for the built-in LSP.
https://marketplace.visualstudio.com/items?itemName=rubocop.vscode-rubocop

Enjoy!
